### ISSUE TITLE
Multi-window support

### ISSUE THREAD

**Issue by jeromelaban** [unoplatform/uno#8341]
Add support for creating multiple windows in the same app, using WinUI provided APIs.

# Notes

- Will require us to start utilizing `SceneDelegate` on iOS/MacCatalyst
  - https://stackoverflow.com/questions/58405393/appdelegate-and-scenedelegate-when-supporting-ios-12-and-13
  - https://manasaprema04.medium.com/scene-delegate-vs-appdelegate-86e22dc17fcb
  - https://medium.com/@kalyan.parise/understanding-scene-delegate-app-delegate-7503d48c5445
  - https://www.appypie.com/app-builder/scene-delegate-app-delegate-xcode-11-ios-13
  - https://www.fivestars.blog/articles/app-delegate-scene-delegate-swiftui/
  - https://www.kairadiagne.com/2019/11/18/adopting-the-new-scene-delegate
- It seems that in WinUI 3, each window is essentially a XAML Island - having `DesktopWindowXamlSource` at the top
- For iOS we will be able to implement `ApplicationView.Title` property and apply it on the ViewController/Scene
- The Window should have its Bounds initialized even in the window constructor - re-enable `AssertInitialWindowSize` on Skia targets

# Resources

- Very useful overview of relationship between various windowing-related classes https://stackoverflow.com/questions/33680631/coreapplicationview-vs-corewindow-vs-applicationview/51585979#51585979
- Additional details on multiple views support https://learn.microsoft.com/en-us/windows/apps/design/layout/show-multiple-views

**Comment by horaciodiez** [comment]
Hi! 
Any updates on this?
Thank you!
Regards

**Comment by horaciodiez** [comment]
> Hi @horaciodiez - what target platforms are you most interested in?

Hi @sasakrsmanovic ! 
I´m most interested in Windows (well, of course) and macOS  (eventually Wasm). 
Having said this, you can avoid the rest of the message :)

We are planning to jump on the WinUI boat now that WinAppSDK 1.1 and MultiWindow support is released.
There are a couple of new projects on the way, and the OS targets are both Windows a macOS. 
We can make Windows mandatory, but if we have a chance to target both it will be welcomed (and so Uno)

In Short, we decided to go to "MultiWindow" in the Desktop / Laptop - LOB Apps environment. 
I may write lot more about this, but i get bored to read myself.

Excuse my poor english.
Thank you!
Regards.

**Comment by notofug** [comment]
Hi, for Wasm builds ; could multi-window support enable the application to present different windows in unique url's at the same time? (ref also https://github.com/unoplatform/uno/issues/5656)
Would it be possible to construct (manually) a HTTP-page that displays the same Wasm-instance in multiple frames, each one viewing different windows of the Wasm-app?
I am wondering how to somehow enable viewing for instance http://wasmapp/status-page and http://wasmapp/config-page etc simultaneously without spawning multipleWasm-app instances in the browser .. all in same browser-tab ; would that be feasible? ;-)

**Comment by renjuashokan** [comment]
Hi, 
I'm also looking to port my current WPF multi-window project to "uno" to make it work on Ubuntu as well. But the lack of <Window> itself makes it impossible to do it.
Will this feature come on anytime soon?

**Comment by FrozDark** [comment]
I don't think we see that till this year :/ Most probably end of the next year

**Comment by MartinZikmund** [comment]
@renjuashokan @FrozDark I would be more optimistic about the availability of this feature, it all depends on the demand. If it is highly requested, it will be prioritized. Our goal also is to implement it properly though - making sure it follows the WinUI implementation as closely as possible so that it is extensible for future

**Comment by wokhan** [comment]
Well, I'm also interested as I only find workarounds I'm not really happy with (using Popup outside of WinUI... as I'm trying right now to use the Window class with WPF as well but not sure it will work with your Skia implementation).
I might also use a global TabView to mimick this multi-window behavior (each tab hosting the "child window" contents) but usual windows would be a way better fit as the goal is to be able to see them all at the same time...).
Thanks!

**Comment by horaciodiez** [comment]
Hi. Happy 2023!
Any good news about this?

**Comment by francoistanguay** [comment]
> ... as I'm trying right now to use the Window class with WPF as well but not sure it will work with your Skia i

Can you reiterate what platforms you're mostly interested in?

For the record, on Windows, Uno isn't present, WinUI runs natively so whatever WinUI supports you can do.

For WASM, how do you foresee Multi-Window working?

Fir macOS, it's pretty straightforward.

**Comment by wokhan** [comment]
Hi, 
For WinUI I've no issue of course, but for WPF/Gtk/WASM which I'm interested in, you can only have one window (the main one) which is blocking any multi window scenario.
I made something pretty simple (while unfinished), Windows are not windows but simply controls loaded on top in the same window, with a fake taskbar (a stack panel where each item matches one of those "windows", allowing to hide/show opened ones).
Basically popups, linked to this "taskbar".
Of course it means you cant go outside the main window boundaries, but at least it works (well, partly, as I didn't finish that yet, I forgot about that while working on other subjects and still have to implement movable popups, which isn't that hard anyway).

**Comment by horaciodiez** [comment]
Yet another relsease and no news on this one. Multi platform - Mono Window. :(

**Comment by FrozDark** [comment]
I think initial support will be in a version 5

**Comment by horaciodiez** [comment]
In @MartinZikmund we trust.

**Comment by horaciodiez** [comment]
> > ... as I'm trying right now to use the Window class with WPF as well but not sure it will work with your Skia i
> 
> Can you reiterate what platforms you're mostly interested in?
> 
> For the record, on Windows, Uno isn't present, WinUI runs natively so whatever WinUI supports you can do.
> 
> For WASM, how do you foresee Multi-Window working?

For Wasm may be something like [this](https://www.egui.rs/#demo)? 😬

**Comment by MartinZikmund** [comment]
@horaciodiez That sounds more like a "in-app window" control/docking control. Could be an interesting feature for https://github.com/unoplatform/uno.toolkit.ui

**Comment by horaciodiez** [comment]
Hi @MartinZikmund !
I believe that it could be like the old MDI interface, but it will be ideal that we can drag the window "off boundaries" the "mdi" container like the sample  attached. (the sample is  a modal dialog, just to show the idea) 
Btw congratulations for the new 4.10 release!
But (always a but) I still need a multi-window platform to build a proper LOB app for Windows and Mac desktop clients. 😬
![Screen-Recording-Made-with-FlexClip-_15_](https://github.com/unoplatform/uno/assets/47337587/2737319d-5422-41b2-ab90-fce5980cdf62)

**Comment by jingliancui** [comment]
Hi All, is it possible for wasm? I plan to choose wasm stack for a product, but users are from enterprise, they are used to use multi browser tabs for their daily tasks.

**Comment by jeromelaban** [comment]
@jingliancui Actual multi-window for webassembly is currently unknown as it would require multi-window shared array buffer sharing and that's not something that the .NET runtime currently supports.

It's possible to create what's been shown in https://github.com/unoplatform/uno/issues/8341#issuecomment-1710476013, though.

**Comment by cyraid** [comment]
Would this include a Docking Manager? It's the one thing that's holding me back from going to Uno Platform from Avalonia. I like how you guys are working with Silk.NET, and just looking at your improvements definitely shows that you guys are pretty serious to be a nice contender to Desktop/Mobile Rapid development.

**Comment by sasakrsmanovic** [comment]
Thanks for that comment @cyraid . We are thinking of Docking Manager in 2025., we are just scoping it out. I'd love to hear your requirements and needs - would you please reach out to sashaATplatform.uno I'd love a 15 min call please.

**Comment by cyraid** [comment]
@sasakrsmanovic Absolutely! I'll send you an email there when I can!

**Comment by MartinZikmund** [comment]
# Repro generator

This issue was classified as not reproducible with a code sample or test.

**Reason:** Multi-window support feature request.

**Comment by horaciodiez** [comment]
> # Repro generator
> This issue was classified as not reproducible with a code sample or test.
> 
> **Reason:** Multi-window support feature request.

Why not?
The example was done with Radzen Dialog.
https://blazor.radzen.com/dialog 
First example.

**Comment by lextm** [comment]
@cyraid even the Avalonia community only has a few docking frameworks that are partially working.

I just ported AvalonDock to Uno Platform and will publish the code soon, so full feature docking support is coming.

<img width="1136" height="784" alt="Image" src="https://github.com/user-attachments/assets/10afde1c-907b-449e-b576-1593c271a2e0" />

@MartinZikmund @sasakrsmanovic Porting AvalonDock indeed revealed some gaps in multi-window support, and I had to resort to native OS interops in a few cases. I will post separate issues as this is a big epic for all, not just Skia desktop.
